blog/_posts/2021-05-18-argocd-private-clusters.md

```diff
@@ -0,0 +1,419 @@
+
+> The steps above should be applied for all the remote cluster you want to manage with the central Argo CD instance. When you have many clusters to manage, this can become quite a task and it can bring additional costs for each LoadBalancer created by your cloud provider. We do have another tutorial available to tackle those issues when [scaling inlets tunnels on Kubernetes](https://inlets.dev/blog/2021/03/15/scaling-inlets.html).
+
+## Connecting the target clusters
+
+Now that the exit-node pods are running, it is time to connect the clients and create a tunnel.
+
+To do so, take all the following steps and execute them against all your target clusters.
+
+Switch your kubectl config to the correct cluster:
+
+As before, I prefer the place the inlets client in a separate namespace:
+
+``` bash
+kubectl create --context gke-eu1 namespace inlets
+```
+
+Now create the deployment for an inlets PRO TCP client:
+
+``` bash
+export LICENSE=$(cat ~/.inlets/LICENSE)
+export TOKEN=<a secure token>
+export DOMAIN=gke-eu1.inlets.example.com
+export TARGET_CTX=gke-eu1
+
+cat <<EOF | kubectl apply --context $TARGET_CTX -f -
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: inlets-client
+  namespace: inlets
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: inlets-client
+  template:
+    metadata:
+      labels:
+        app: inlets-client
+    spec:
+      containers:
+      - name: inlets-client
+        image: ghcr.io/inlets/inlets-pro:0.8.4
+        imagePullPolicy: IfNotPresent
+        command: ["inlets-pro"]
+        args:
+        - "tcp"
+        - "client"
+        - "--url=wss://${DOMAIN}:8123/connect"
+        - "--upstream=kubernetes.default.svc"
+        - "--port=443"
+        - "--token=${TOKEN}"
+        - "--license=${LICENSE}"
+EOF
+```
+
+> This example is creates a client with a single yaml configuration. There is also an Helm [chart](https://github.com/inlets/inlets-pro/tree/master/chart/inlets-pro-client) available for advanced usages.
+
+## Adding the cluster to Argo CD
+
+With the tunnel up and running, we can now register the private cluster to Argo CD.
+
+Typically you could add a cluster using the Argo CD CLI command `argocd cluster add CONTEXTNAME`, where the context name is a context available in your current kubectl config. That command installs a ServiceAccount (`argocd-manager`) into the `kube-system` namespace of that kubectl context, and binds the service account to an admin-level ClusterRole.
+
+Unfortunately, adding a cluster this way will fail in our scenario. Besides creating the service account in the target cluster, the command will also try to register the cluster in Argo CD with the endpoint in your context and will validate if Argo CD can communicate with the API service. It should be obvious the latter will not succeed.
+
+Luckily for us, we can configure everything in a declarative way.
```

Comment:
We can simulate the steps that the CLI takes to onboard a new cluster. (Insert brief summary of the steps that will follow)